{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "68fd795b-6ba2-4f38-9d31-ffa6007dacbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'California': ['Sacramento', 'San Francisco', 'Lake Tahoe', 'San Diego'], 'Ohio': ['Cleveland', 'Springville']}\n"
     ]
    }
   ],
   "source": [
    "# how to store more than one item in a dictionary spot\n",
    "cities_in_states = {}\n",
    "# only one item\n",
    "cities_in_states['California'] = 'Sacramento'\n",
    "cities_in_states['California'] = ['Sacramento', 'San Francisco', 'Lake Tahoe']\n",
    "cities_in_states['California'].append('San Diego')\n",
    "cities_in_states['Ohio'] = ['Cleveland']\n",
    "cities_in_states['Ohio'].append('Springville')\n",
    "\n",
    "print(cities_in_states)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d33fd657-30e4-47a7-92d7-a9b5304c0f31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'apple': {'color': 'red', 'size': 'medium', 'seeds': True, 'shape': 'round'}, 'orange': {'color': 'orange', 'shape': 'round', 'size': 'medium', 'taste': 'sweet', 'seeds': True, 'rollability': 'excellent'}, 'grape_t': {'color': 'purple', 'shape': 'round', 'size': 'small', 'seeds': True}, 'grape_g': {'color': 'green', 'shape': 'round', 'size': 'small', 'seeds': False}, 'pear': {'color': 'green', 'shape': 'pear', 'size': 'medium', 'seeds': True}}\n",
      "Fruit apple has a color of red\n",
      "Fruit orange has a color of orange\n",
      "Fruit grape_t has a color of purple\n",
      "Fruit grape_g has a color of green\n",
      "Fruit pear has a color of green\n",
      "==\n",
      "Name\tsize\tshape\tcolor\n",
      "apple\tmedium\tround\tred\n",
      "orange\tmedium\tround\torange\n",
      "grape_t\tsmall\tround\tpurple\n",
      "grape_g\tsmall\tround\tgreen\n",
      "pear\tmedium\tpear\tgreen\n"
     ]
    }
   ],
   "source": [
    "import pprint\n",
    "fruit = {}\n",
    "fruit['apple'] = {'color': 'red', 'size': 'medium', 'seeds': True}\n",
    "fruit['apple']['shape'] = 'round'\n",
    "fruit['orange'] = {}\n",
    "fruit['orange']['color'] = 'orange'\n",
    "fruit['orange']['shape'] = 'round'\n",
    "fruit['orange']['size'] = 'medium'\n",
    "fruit['orange']['taste'] = 'sweet'\n",
    "fruit['orange']['seeds'] = True\n",
    "fruit['orange']['rollability'] = 'excellent'\n",
    "fruit['grape_t'] = {'color': 'purple', 'shape': 'round', 'size': 'small', 'seeds': True}\n",
    "fruit['grape_g'] = {'color': 'green',  'shape': 'round', 'size': 'small', 'seeds': False}\n",
    "fruit['pear'] = {'color': 'green',  'shape': 'pear', 'size': 'medium', 'seeds': True}\n",
    "\n",
    "#fruit['grapes'] = ['purple']\n",
    "print(fruit)\n",
    "for f in fruit:\n",
    "    #print(f)\n",
    "    #print(fruit[f]['color'])\n",
    "    print(f'Fruit {f} has a color of {fruit[f][\"color\"]}')\n",
    "\n",
    "print(\"==\")\n",
    "properties = sorted(['color', 'shape', 'size'],reverse=True)\n",
    "print(\"\\t\".join(['Name']+properties))\n",
    "for f in fruit:\n",
    "#    row = [f, fruit[f]['color'],fruit[f]['shape'], fruit[f]['size']]\n",
    "    row = [f]\n",
    "    for prop in properties:\n",
    "        row.append(fruit[f][prop])\n",
    "    print(\"\\t\".join(row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9db557fb-912b-4846-8058-eb495a282522",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'apple': [{'id': 'app012', 'wt': 10.2, 'size': 5}, {'id': 'app013', 'wt': 9.8, 'size': 4.2}, {'id': 'app014', 'wt': 6.7, 'size': 3.2}, {'id': 'app017', 'wt': 7.1, 'size': 1.2}, {'id': 'app016', 'wt': 3.2, 'size': 2.2}], 'orange': [{'id': 'or01', 'wt': 4.7, 'size': 3.1}]}\n",
      "Average apple weight is 7.4 out of 5 apples\n",
      "Average apple weight is 7.4 out of 5 apple\n",
      "Average orange weight is 4.7 out of 1 orange\n"
     ]
    }
   ],
   "source": [
    "fruit = {}\n",
    "# barrel with multiple types of fruit, and a fruit can occur more than once\n",
    "#type='apple'\n",
    "fruit['apple'] = []\n",
    "fruit['apple'].append({'id': 'app012', 'wt': 10.2, 'size': 5})\n",
    "fruit['apple'].append({'id': 'app013', 'wt': 9.8, 'size': 4.2})\n",
    "fruit['apple'].append({'id': 'app014', 'wt': 6.7, 'size': 3.2})\n",
    "fruit['apple'].append({'id': 'app017', 'wt': 7.1, 'size': 1.2})\n",
    "fruit['apple'].append({'id': 'app016', 'wt': 3.2, 'size': 2.2})\n",
    "# apple = {'id': 'app016', 'wt': 3.2, 'size': 2.2}\n",
    "# \n",
    "fruit['orange'] = []\n",
    "fruit['orange'].append({'id': 'or01', 'wt': 4.7, 'size': 3.1})\n",
    "\n",
    "print(fruit)\n",
    "# average weight of an apple?\n",
    "total_weight = 0\n",
    "total_apples = 0\n",
    "for apple in fruit['apple']:\n",
    "    total_weight += apple['wt']\n",
    "    total_apples += 1\n",
    "print(f'Average apple weight is {total_weight / total_apples} out of {total_apples} apples')\n",
    "for type in fruit:\n",
    "    total = 0\n",
    "    weight = 0\n",
    "    for f in fruit[type]:\n",
    "        weight += f['wt']\n",
    "        total += 1\n",
    "    print(f'Average {type} weight is {weight / total} out of {total} {type}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "af2c4693-b23b-4577-a5c6-c2ea518cbf9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rice filename is rice_random_exons.bed\n",
      "Total length of exons is 370855 bp (370.855 kb)\n",
      "Total length of exons is 370855 bp (370.855 kb)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "riceurl=\"https://raw.githubusercontent.com/biodataprog/GEN220/master/data/rice_random_exons.bed\"\n",
    "ricefile=os.path.basename(riceurl)\n",
    "print(\"Rice filename is {}\".format(ricefile))\n",
    "if not os.path.exists(ricefile):\n",
    "    os.system(\"curl -O {}\".format(riceurl))\n",
    "total_exon_length = 0\n",
    "with open(ricefile,\"r\") as fh:\n",
    "    for bedline in fh:\n",
    "#        bedline = bedline.strip()\n",
    "        bedarray = bedline.strip().split(\"\\t\")\n",
    "        exonlen = abs(int(bedarray[2])-int(bedarray[1]))+1\n",
    "#        print(\"exon length is {}\".format(exonlen))\n",
    "        total_exon_length += exonlen\n",
    "print(\"Total length of exons is {} bp ({} kb)\".format(total_exon_length,total_exon_length/1000))\n",
    "print(f\"Total length of exons is {total_exon_length} bp ({total_exon_length/1000} kb)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "97dfc2f5-ab3e-481b-a8e8-c247dcc40282",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species filename is species.csv\n",
      "===\n",
      "now ordered by abundance using the value stored in the dictionary\n",
      "Dipodomys\t4\n",
      "Chaetodipus\t4\n",
      "Reithrodontomys\t4\n",
      "Sigmodon\t4\n",
      "Onychomys\t3\n",
      "Pipilo\t3\n",
      "Peromyscus\t3\n",
      "Crotalus\t2\n",
      "Cnemidophorus\t2\n",
      "Neotoma\t2\n",
      "Perognathus\t2\n",
      "Sceloporus\t2\n",
      "Spermophilus\t2\n",
      "Amphispiza\t1\n",
      "Ammospermophilus\t1\n",
      "Ammodramus\t1\n",
      "Baiomys\t1\n",
      "Campylorhynchus\t1\n",
      "Calamospiza\t1\n",
      "Callipepla\t1\n",
      "Eumeces\t1\n",
      "Gambelia\t1\n",
      "Pooecetes\t1\n",
      "Sylvilagus\t1\n",
      "Spizella\t1\n",
      "Lizard\t1\n",
      "Rodent\t1\n",
      "Sparrow\t1\n",
      "Zonotrichia\t1\n",
      "Zenaida\t1\n"
     ]
    }
   ],
   "source": [
    "import os, csv, sys\n",
    "speciesurl=\"https://raw.githubusercontent.com/biodataprog/GEN220_2025_classexamples/refs/heads/main/python/species.csv\"\n",
    "spfile=os.path.basename(speciesurl)\n",
    "print(\"Species filename is {}\".format(spfile))\n",
    "if not os.path.exists(spfile):\n",
    "    os.system(\"curl -O {}\".format(speciesurl))\n",
    "\n",
    "genera = {} # this is a dictionary\n",
    "with open(spfile,\"r\") as fh:\n",
    "    speciesreader = csv.reader(fh, delimiter=',')\n",
    "    header = next(speciesreader)\n",
    "    for row in speciesreader:\n",
    "        genus = row[1]\n",
    "        if len(genus) == 0:\n",
    "            continue\n",
    "        if genus in genera:\n",
    "            genera[genus] += 1\n",
    "        else:\n",
    "            genera[genus] = 1\n",
    "\n",
    "generaout = csv.writer(sys.stdout,delimiter=\"\\t\")\n",
    "#for genus in sorted(genera):\n",
    "#    generaout.writerow([genus,genera[genus]])\n",
    "    #print(\"{}\\t{}\".format(genus,genera[genus]))\n",
    "print(\"===\")\n",
    "print(\"now ordered by abundance using the value stored in the dictionary\")\n",
    "generaout = csv.writer(sys.stdout,delimiter=\"\\t\")\n",
    "for genus in sorted(genera.keys(), key=lambda x: genera[x],reverse=True):\n",
    "    generaout.writerow([genus,genera[genus]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "6fdea333-2e0f-4620-9c7b-e94dbee6dac6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filename is S_cerevisiae.ORFs.fasta\n",
      "1.\n",
      "There are 6713 sequences in the file\n",
      "2.\n",
      "The distribution of first codons is:\n",
      "ATG => 6706 (99.9%)\n",
      "TAT => 1 (0.0%)\n",
      "ATA => 6 (0.1%)\n",
      "The distribution of last codons is:\n",
      "TAA => 3160 (47.1%)\n",
      "TGA => 1995 (29.7%)\n",
      "TAG => 1558 (23.2%)\n",
      "3.\n",
      "There are 3042 genes on the Watson (+) strand\n",
      "          3023 genes on the Crick  (-) strand\n"
     ]
    }
   ],
   "source": [
    "import itertools, sys, re\n",
    "\n",
    "# define what a header looks like in FASTA format\n",
    "def isheader(line):\n",
    "    return line[0] == '>'\n",
    "\n",
    "def aspairs(filename):\n",
    "    seq_id = ''\n",
    "    sequence = ''\n",
    "    with open(filename,\"r\") as f:\n",
    "        for header,group in itertools.groupby(f, isheader):\n",
    "            if header:\n",
    "                line = next(group)\n",
    "                seq_id = line[1:].split()[0]\n",
    "            else:\n",
    "                sequence = ''.join(line.strip() for line in group)\n",
    "                yield seq_id, sequence\n",
    "\n",
    "# you could use this or you can download this from here instead\n",
    "# https://github.com/biodataprog/GEN220_data/raw/main/data/S_cerevisiae_ORFs.fasta\n",
    "yeastorfs=\"https://github.com/biodataprog/GEN220_data/raw/main/data/S_cerevisiae_ORFs.fasta.gz\"\n",
    "filename=os.path.basename(yeastorfs)\n",
    "filename=\"S_cerevisiae.ORFs.fasta\"\n",
    "#if not os.path.exists(filename):\n",
    "#    os.system(\"curl -L -O {}\".format(yeastorfs))\n",
    "#    os.system(\"gunzip {}\".format(yeastorfs))\n",
    "#filename=\"/bigdata/gen220/shared/data/S_cerevisiae_ORFs.fasta\"\n",
    "\n",
    "print(\"filename is {}\".format(filename))\n",
    "seqs = dict(aspairs(filename))\n",
    "first_codon_dict = {}\n",
    "last_codon_dict  = {}\n",
    "sequence_count = 0\n",
    "strand_count = {}\n",
    "i = 0\n",
    "for seqname in seqs:\n",
    " #   print(seqname)\n",
    "    sequence_count += 1\n",
    "    firstcodon = seqs[seqname][0:3]\n",
    "    #print(\"length of sequence = \",len(seqs[seqname]))\n",
    "    lastcodon = seqs[seqname][-3:]\n",
    "#    count = first_codon_dict.get(firstcodon,0)\n",
    "#    first_codon_dict.set(firstcodon,count+1)\n",
    "    if firstcodon in first_codon_dict:\n",
    "         first_codon_dict[firstcodon] +=1\n",
    "    else:\n",
    "        first_codon_dict[firstcodon] = 1\n",
    "\n",
    "    if lastcodon in last_codon_dict:\n",
    "        last_codon_dict[lastcodon] +=1\n",
    "    else:\n",
    "        last_codon_dict[lastcodon] = 1\n",
    "#       print(firstcodon,lastcodon)\n",
    "    last_char = seqname[-1]\n",
    "#       print(last_char, \" in \", seqname)\n",
    "    if last_char in strand_count:\n",
    "        strand_count[last_char] += 1\n",
    "    else:\n",
    "        strand_count[last_char] = 1\n",
    "print(\"1.\")\n",
    "print(\"There are %d sequences in the file\"%(sequence_count))\n",
    "\n",
    "print(\"2.\")\n",
    "print(\"The distribution of first codons is:\")\n",
    "\n",
    "for codon in first_codon_dict:\n",
    "    print(\"%s => %d (%.1f%%)\" % (codon, first_codon_dict[codon],\n",
    "                                 100.0 * first_codon_dict[codon] / sequence_count))\n",
    "\n",
    "\n",
    "print(\"The distribution of last codons is:\")\n",
    "for codon in last_codon_dict:\n",
    "    print(\"%s => %d (%.1f%%)\" % (codon, last_codon_dict[codon],\n",
    "                                 100.0 * last_codon_dict[codon] / sequence_count))\n",
    "\n",
    "print(\"3.\")\n",
    "print(\"There are %s genes on the Watson (+) strand\"%(strand_count['W']))\n",
    "print(\"          %s genes on the Crick  (-) strand\"%(strand_count['C']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
